src/skagent/ann.py

```diff
@@ -657,51 +685,44 @@ def train_block_value_and_policy_nn(
         The value network to train
     inputs : Grid
         Input grid containing states and shocks
-    policy_loss_function : callable
-        Loss function for policy training (takes decision_function, input_grid)
-    value_loss_function : callable
-        Loss function for value training (takes value_function, input_grid)
+    loss_function : callable
+        Unified loss function that takes (value_function, decision_function, input_grid)
+        and trains both networks simultaneously
     epochs : int, optional
-        Number of training epochs, by default 50
+        Number of training epochs, by default 250
 
     Returns
     -------
     tuple
         (trained_policy_nn, trained_value_nn)
     """
-    # to change
-    # criterion = torch.nn.MSELoss()
-    policy_optimizer = torch.optim.Adam(
-        block_policy_nn.parameters(), lr=0.01
-    )  # Using Adam
-    value_optimizer = torch.optim.Adam(
-        block_value_nn.parameters(), lr=0.01
-    )  # Using Adam
+    # Joint optimizer for both networks
+    joint_optimizer = torch.optim.Adam(
+        list(block_policy_nn.parameters()) + list(block_value_nn.parameters()), lr=0.01
+    )
 
     for epoch in range(epochs):
-        # Train policy network
-        policy_optimizer.zero_grad()
-        policy_loss = aggregate_net_loss(
-            inputs, block_policy_nn.get_decision_function(), policy_loss_function
-        )
-        policy_loss.backward()
-        policy_optimizer.step()
+        joint_optimizer.zero_grad()
+
+        # Single loss function trains both networks
+        # Follow exact same pattern as train_block_policy_nn using aggregate_net_loss
+        running_loss = 0.0
+
+        # Create adapted loss function that captures the value function
+        value_function = block_value_nn.get_value_function()
 
-        # Train value network
-        value_optimizer.zero_grad()
-        value_loss = aggregate_net_loss(
-            inputs, block_value_nn.get_value_function(), value_loss_function
+        def adapted_loss_function(df, inputs):
+            return loss_function(value_function, df, inputs)
+
+        joint_loss = aggregate_net_loss(
+            inputs, block_policy_nn.get_decision_function(), adapted_loss_function
```

Comment:
This seems to be hacking around what was intended to be a unified type signature for loss functions.

If that abstraction is being broken it means the design needs better thinking.